lib/phoenix/code_reloader/server.ex

```diff
@@ -129,25 +138,37 @@ defmodule Phoenix.CodeReloader.Server do
     end
   end
 
-  defp mix_compile({:module, Mix.Task}, compilers) do
-    if Mix.Project.umbrella? do
-      Enum.each Mix.Dep.Umbrella.cached, fn dep ->
-        Mix.Dep.in_dependency(dep, fn _ ->
-          mix_compile_unless_stale_config(compilers)
-        end)
-      end
-    else
-      mix_compile_unless_stale_config(compilers)
-      :ok
-    end
+  defp mix_compile({:module, Mix.Task}, compilers, apps_to_reload) do
+    reloaded = mix_compile_deps(Mix.Dep.cached, apps_to_reload, compilers)
+
+    mix_compile_project(Mix.Project.config()[:app], apps_to_reload, reloaded, compilers)
   end
-  defp mix_compile({:error, _reason}, _) do
+  defp mix_compile({:error, _reason}, _, _) do
     raise "the Code Reloader is enabled but Mix is not available. If you want to " <>
           "use the Code Reloader in production or inside an escript, you must add " <>
           ":mix to your applications list. Otherwise, you must disable code reloading " <>
           "in such environments"
   end
 
+  defp mix_compile_deps(deps, apps_to_reload, compilers) do
+    deps
+    |> Enum.filter(&(&1.app in apps_to_reload))
+    |> Enum.map(fn dep ->
+      Mix.Dep.in_dependency dep, fn _ ->
+        mix_compile_unless_stale_config(compilers)
+      end
```

Comment:
Are you sure this `map` is returning application names? Maybe we should explicitly return the app name? Also, I think this pipeline would be clearer as a for comprehension.

Reply:
This is a very good point, it is not working as expected.

In fact this check for already compiled apps does nothing and I can clean it up. In umbrella project there is no "project app", and in non-umbrella it is there but not in deps. It's not going to happen that we need to check if the main app was already compiled.